### PR TITLE
fix nan/inf loss

### LINKED ISSUE #186: Nan loss during first epoch in custom dataset training
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Training

### Bug

Congratz on the release!

I start a training with yolov8m model with 1280 input size, 2 batch, for 10 epochs on my custom dataset with 1 target class. During %10 of the first epoch, loss becomes nan and map becomes 0.

My GPU is 1x2080ti and 4.1GB of 11GB vram is used during training.

### Environment

- Python 3.10.8
- Ubuntu 20.04
- YOLOv8 8.0.0
- torchvision 0.13.1 (py310_cu113)
- pytorch 1.12.1 (py3.10_cuda11.3_cudnn8.3.2_0)

### Minimal Reproducible Example

_No response_

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### LINKED ISSUE #280: nan report in box_class cls_class and dfl_loss when train custom dataset
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Training

### Bug

Hello, I am newbie in computer vision and I just started to try the new version yolov8 and I get some error when take the result.
I seem like something wrong but I don't know how to fix it. Can you give me some suggest?

bug

### Environment

-YOLOv8n
-CUDA: 11.6
-Ultralytics YOLOv8.0.4
-OS: Windows 10

### Minimal Reproducible Example

_No response_

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### LINKED ISSUE #461: All losses went to nan and precision & recall went low & static
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

What inference should I take from this? 

I ran YOLO v8 for 300 epochs but after 7 epochs all loss values went to nan, and precision & recall went static to very low. 

image

### Additional

_No response_

### PR DESCRIPTION
**Pull request by Laughing-q**
@AyushExel @glenn-jocher this PR fixed the nan loss issue. And the cause is that the `target_scores_sum` we're using in loss calculation could be 0 if there're no objects in targets(empty labels, background).

test command:
```bash
yolo detect train model=yolov8n.pt data=custom.yaml imgsz=640 rect=True
```
before fix(losses are nan and mAP stays 0.15):
pic-selected-230119-1453-35
after fix:
pic-selected-230119-1521-59

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced stability in loss calculation during model training.

### 📊 Key Changes
- Modified the calculation of `target_scores_sum` by ensuring it's never less than 1.

### 🎯 Purpose & Impact
- 🛠 Prevents division by zero errors when calculating loss, increasing the robustness of training.
- ⚖️ Contributors and users of the Ultralytics framework can expect more stable training sessions, particularly in edge cases with sparse detection targets.

### CODE REVIEW

**Comment by hdnh2006** [comment]
@Laughing-q I am still getting nan in my training. It seems for validation is solved:
image

After running `pip install --upgrade ultralytics` I get the following:
```
Looking in indexes: https://pypi.org/simple, https://pypi.ngc.nvidia.com
Requirement already satisfied: ultralytics in ********/.virtualenvs/ultralytics/lib/python3.8/site-packages (8.0.10)
```

**Comment by AyushExel** [comment]
@hdnh2006 wait for a while till this gets merged. We'll release the updated package later today

**Comment by CoderYiFei** [comment]
well done @Laughing-q 
I always get nan value in training dataset , and valid dataset.
thanks very much.
I am waiting for the master' s code, that fixed the nan value bug.

**Comment by glenn-jocher** [comment]
@Laughing-q do you know what a typical value of `target_scores_sum` is? Could this change affect COCO training results?

Should we add a smaller value for a protected divide, i.e. `x / (target_scores_sum + eps)` where eps might be something like  `1-e6`

**Comment by Laughing-q** [comment]
@glenn-jocher oh yes, I just want to tell you this.
image
